### PR TITLE
Implement optional thread interrupt on coroutine cancellation (#57)

### PR DESCRIPTION
**Pull request by jxdabc**
See issue #57 for details

Signed-off-by: Trol <jiaoxiaodong@xiaomi.com>

### CODE REVIEW

File: kotlinx-coroutines-core/jvm/src/CoroutineInterruptible.kt [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+                    s === Interrupting || s === Interrupted -> return
+                    else -> throw IllegalStateException("unknown state")
+                }
+            }
+        }
+
+        private fun initThread() {
+            val thread = Thread.currentThread()
+            state.loop { s ->
+                when {
```
Comment (taralx) [inline kotlinx-coroutines-core/jvm/src/CoroutineInterruptible.kt:105]: Perhaps `when (s) {`?

File: kotlinx-coroutines-core/jvm/src/CoroutineInterruptible.kt [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+                    else -> throw IllegalStateException("unknown state")
+                }
+            }
+        }
+
+        private inner class CancelHandler : CompletionHandler {
+            override fun invoke(cause: Throwable?) {
+                state.loop { s ->
+                    when {
+                        s === Init || (s is Working && s.thread === null) -> {
```
Comment (taralx) [inline kotlinx-coroutines-core/jvm/src/CoroutineInterruptible.kt:119]: If you swap initThread and initInvokeOnCancel then this can't happen, right?

Comment (jxdabc) [reply kotlinx-coroutines-core/jvm/src/CoroutineInterruptible.kt:119]: seems a great idea, I will check it

**Comment by jxdabc** [comment]
@qwwdfsad 
This feature is of great use at least on Android as I could see. Please see ceses in #57 .
There is still another non-intrusive approach to this feature we are currently using, though not as easy to use as this one. I'll post another PR including that approach. Hope that would help.

**Comment by jxdabc** [comment]
@qwwdfsad 
Hoping you'd have a look at: 

#1934
(non-intrusive) Implement optional thread interrupt on coroutine cancellation (#57)

**Comment by elizarov** [comment]
Superseded by #1934

**Pull request closed without merging**